Autocomplete for item search, static item name array

Loraykin opened the pull request:
The datasource for the item names needs addressing. Currently static array, should be a cached API result or local db that gets updated periodically.

Loraykin committed:
Autocomplete for item search, static item name array

Mandalorian007 commented on .gitignore:
```diff
@@ -1,2 +1,4 @@
-/node_modules/
\ No newline at end of file
+/node_modules/
+config.json
```
OMG I totally forgot to add config.json here. can you also end an entry for `test-config.json`?

Loraykin replied:
Done

Mandalorian007 commented on commands/item.js:
```diff
@@ -38,24 +40,60 @@ Examples:
-  const itemName = temp.join(' ');
+  let itemName = temp.join(' ');
+
+  let candidates = resolveItemName(itemName); // check item name against local static array 
```
As a general rule of thumb shift your comments to the line above rather than at the end, but love the detailing you are doing in your comments

Loraykin replied:
I guess this comes down to personal preference. I find it useful to have two levels for comments, one pertaining to all code beneath it (until next full-line comment, closure, paragraph...), one limited to the actual line. It's also a way to keep code from blowing up too much, and make sure you do not pass on commenting just because it breaks the code flow. I've changed it somewhat though in the next commit.

Mandalorian007 commented on commands/item.js:
```diff
@@ -38,24 +40,60 @@ Examples:
-  const itemName = temp.join(' ');
+  let itemName = temp.join(' ');
+
+  let candidates = resolveItemName(itemName); // check item name against local static array 
```
Candidates can be a const here I believe

Loraykin replied:
Would have been true, now I reassign it for sorting.

Mandalorian007 commented on commands/item.js:
```diff
@@ -38,24 +40,60 @@ Examples:
     + param;
   const response = await fetch(url);
-  return response.ok ? printItem(await response.json()) : 'Unable to locate item.';
+  return response.ok ? printItem(await response.json()) : new Discord.RichEmbed()
```
This looks a bit crazy around the ternary if statement. Can we move the richEmbed assembly to a helper method?

Loraykin replied:
Moved to methods in the same file

Mandalorian007 commented on commands/item.js:
```diff
@@ -38,24 +40,60 @@ Examples:
+    .setThumbnail('attachment://noidea.png');
 });
 
+/// match an item name against the static array of available item names
```
I'd love a TODO comment for updating this to be dynamic and cached

Loraykin replied:
Done, added to method

Mandalorian007 commented on commands/item.js:
```diff
@@ -38,24 +40,60 @@ Examples:
+  try {
+    let module = require('../data/itemNames.js');
+    let itemNames = module.itemNames;
+    return itemNames.filter(i => i.toLowerCase().includes(name.toLowerCase()));
```
I think it would be pretty low lift to apply a sort here that anything that "starts with" the partial name would rise to the top of the suggestions. Might help make the suggestions slightly better

Loraykin replied:
Good point. Was on my mind, too, but then forgot about it again

Loraykin replied:
Implemented

Mandalorian007 commented on commands/item.js:
```diff
@@ -38,24 +40,60 @@ Examples:
 });
 
+/// match an item name against the static array of available item names
+const resolveItemName = (name) => {
```
ResolveItemName sounds very definitive for returning a candidate list maybe a rename on this

Loraykin replied:
Renamed to matchItemName

Mandalorian007 reviewed (commented):
Overall this is looking really fantastic and I love the approach

Loraykin committed:
Sort suggestions by significance

Mandalorian007 commented on commands/item.js:
```diff
@@ -38,24 +58,71 @@ Examples:
+    else if (candidates.length > 1) {
+      // multiple matches, suggest some candidates
+      // prioritise items which start with the input term
+      candidates = candidates.sort((a, b) => a.toLowerCase().indexOf(itemName.toLowerCase())
```
This is a very clean way of implementing this I love it :)

Mandalorian007 reviewed (approved):
Approved this is looking fantastic! I'm comfortable rolling out this first pass if you are.

Loraykin committed:
Autocomplete for orbs (copy/paste from item.js)
Using pure URL instead of attachment for message thumbnail.
'Multiple results'/'no results' messages moved to commands/_helper.js

Loraykin commented on commands/_helper.js:
```diff
@@ -1,3 +1,19 @@
+  new Discord.RichEmbed()
+    .setTitle(`${searchCategory} '${searchTerm}' not found`)
+    .addField('Please check your input', '\u200b')
+    .setThumbnail('https://questland-discord-bot.cfapps.io/noidea.png');
```
Where do you want this? In my usual coding, I would create a new folder "Helpers", and a new file "messageHelper.js" in a case like this.

Mandalorian007 replied:
Yeah I am good with a helper directory here it seems like a solid choice to me

Mandalorian007 replied:
You can leave the _helper file there since that is specific to how I have asynchronous commands managed

Loraykin commented on commands/orb.js:
```diff
@@ -28,16 +30,65 @@ Examples:
   const response = await fetch(url);
-  return response.ok ? printOrb(await response.json()) : 'Unable to locate orb.';
+  
+  return response.ok ? printOrb(await response.json()) : noResultFoundMessage(orbName, 'Orb');
```
The 'no result' currently throws an error for orbs (unexpected end of JSON input), take a look at the public API code to solve

Mandalorian007 replied:
fixed

Mandalorian007 commented on commands/orb.js:
```diff
@@ -1,6 +1,8 @@
 const Discord = require("discord.js");
 const { asyncHandler } = require("./_helper");
+const { multipleResultsFoundMessage } = require("./_helper")
+const { noResultFoundMessage } = require("./_helper")
```
Minor thing. You can add a comma after multipleResultsFoundMessage and simply have noResultFoundMessage to keep this at 1 line of imports per file. Not too concerned about this

Loraykin replied:
ok

Loraykin committed:
Moved messages to own helper
